moon/controller.py

```diff
@@ -479,6 +483,48 @@ def on_scan(self, data):
         before_robot.sort();
         self.scan_distance_to_obstacle = median(before_robot[:self.scan_min_window])
 
+        # unless we are already inside unsafe area, augment nearest obstacles with virtual fence values
+        if self.unsafe_location is not None and not self.unsafe_location.contains(Point(self.xyz[0], self.xyz[1])):
+            # calculate position in front of and to the left and right of the robot in the direction of its heading
+            v = np.asmatrix(np.asarray([self.xyz[0], self.xyz[1], 1]))
+            c, s = np.cos(self.steering_angle + self.yaw), np.sin(self.steering_angle + self.yaw)
+            R = np.asmatrix(np.array(((c, -s, 0), (s, c, 0), (0, 0, 1))))
+
+            T = np.asmatrix(np.array(((1, 0, +6),(0, 1, 0),(0, 0, 1))))
+            pos =  np.dot(R, np.dot(T, np.dot(R.I, v.T)))
+            ls = LineString([(self.xyz[0], self.xyz[1]), (float(pos[0]), float(pos[1]))])
+            intr = self.unsafe_location.intersection(ls)
+            if not intr.is_empty:
+                d = int(1000*(ls.length - intr.length))
+                if self.debug:
+                    print(self.sim_time, self.robot_name, "Virtual fence: decreasing front distance to %d" % d)
+                self.scan_distance_to_obstacle = min(self.scan_distance_to_obstacle, d)
+
+            left_angle = 0.787878
```

Comment:
this looks like some magic number

Reply:
It's actually derived from the numbers a few lines up. If there is an obstacle ahead, we are trying to decide if it's better to avoid it by turning left or right. This is done by measuring min distance across IDEAL_SAMPLE_SIZE lidar beams to the left and to the right of which way the steering points to. (The IDEAL_SAMPLE_SIZE is 30 by default but if the steering is making sharp turn, there may be fewer lidar beams left in that direction). For virtual fence, we don't make the calculations for each angle representing each lidar beam, rather we just make the calculation for the most extreme one. If the angle difference between two consecutive beams is  0.0262626260519, this is 30x that. I can probably replace it by a formula for better readability.